airflow/timetables/datasets.py

```diff
@@ -0,0 +1,90 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+import collections.abc
+import typing
+
+import attrs
+
+from airflow.datasets import Dataset
+from airflow.exceptions import AirflowTimetableInvalid
+from airflow.timetables.simple import DatasetTriggeredTimetable
+from airflow.utils.types import DagRunType
+
+if typing.TYPE_CHECKING:
+    import pendulum
+
+    from airflow.timetables.base import DagRunInfo, DataInterval, TimeRestriction, Timetable
+
+
+class DatasetTimetable(DatasetTriggeredTimetable):
```

Comment:
We do need to come up with a better name.  Maybe start a thread on slack for some brainstorming?

There should be something to indicate that it is dataset plus schedule... unless.... unless ... actually maybe we should just update DatasetTriggeredTimetable so that it works like this?  @uranusjr ?

Reply:
Yeah. I would be also for just merging it with DatasetTriggeredTimetable (where default behaviour on dataset dependencies would be time = None. I think there is no need to muliply the classes. As the docstring on DatasetTriggeredTimetable states - it should not be used anywhere by the users in DAGs, so we should be able to change the name easily.

My proposals for that combined classs would be `DatasetOrTimeTiggeredTimetable`.

Reply:
I liked `DatasetOrTimeTiggeredTimetable` better as the purpose is clear from this name.

Reply:
I am just reading about TimeTable and have not really used it before.

Looking at the current codebase, would like to understand the below point?
I am wondering if it's wise to have called the existing class as `DatasetTriggeredTimetable` - I am thinking that since the trigger is not actually based on time, should it still be a TimeTable (meaning should it contain TimeTable in its name DatasetTriggeredTimetable)?

If we feel it wise, then I have a +1 to merging these 2 with default `time=None` and also +1 to `DatasetOrTimeTiggeredTimetable`.
Otherwise, if we consider renaming the existing class and creating another class for the OR, then can rethink of the name to provide inputs here

Reply:
I am afraid that  ship has sailed (TimeTable). It's original sin of using TimeTable where we have not imagined ther might be other triggers. 

But we could potentially consider making an Alias and rename the base class as well - possibly that's worth considering. Unfortunately Trigger has alredy been taken for something else (this would be probably best name). 

Maybe just `Schedule` - taking into account that the canonical name of the parameter is "schedule' and:

```
ScheduleArg = Union[ArgNotSet, ScheduleInterval, Timetable, Collection["Dataset"]]
```

While schedule is also time-bound, it's far less than TimeTable:

```python
DatasetOrTimeSchedule(CustomSchedule):
```

That would be nice if we do it this way I think,

Reply:
But that would have a bit of a ripple-effect on a number of classes I am afraid.

Reply:
Changed this: https://github.com/apache/airflow/pull/36710/commits/9c3abb7b2e2dad2fd7122db535a364b2433087c1